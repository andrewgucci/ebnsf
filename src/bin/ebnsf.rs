use ebnsf::parse_ebnf;

use std::path::PathBuf;

fn main() {
    let ebnf = std::fs::read_to_string("test/bnf.ebnf").unwrap();

    let diagram = parse_ebnf(&ebnf).unwrap();

    let output = PathBuf::from("test.svg");

    std::fs::write(&output, diagram.to_string().into_bytes()).unwrap();
}
