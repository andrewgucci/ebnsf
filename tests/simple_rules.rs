use ebnsf::parse_ebnf;

mod common;
use common::save_svg;

#[test]
fn simple_choice() {
    let rule = "<rule> ::= <foo> | <bar>";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("simple_choice", &svg);

    insta::assert_snapshot!(svg);
}

#[test]
fn simple_optional() {
    let rule = "<rule> ::= <foo>?";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("simple_optional", &svg);

    insta::assert_snapshot!(svg);
}

#[test]
fn simple_star() {
    let rule = "<rule> ::= <foo>*";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("simple_star", &svg);

    insta::assert_snapshot!(svg);
}

#[test]
fn simple_plus() {
    let rule = "<rule> ::= <foo>+";

    let svg = parse_ebnf(rule).expect("valid grammar").to_string();

    save_svg("simple_plus", &svg);

    insta::assert_snapshot!(svg);
}
