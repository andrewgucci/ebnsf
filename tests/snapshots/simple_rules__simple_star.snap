---
source: tests/simple_rules.rs
expression: svg
---
<svg class="railroad" viewBox="0 0 234 68" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<style type="text/css">
svg.railroad {
background-color: hsl(30, 20%, 95%);
background-size: 15px 15px;
background-image: linear-gradient(to right, rgba(30, 30, 30, .05) 1px, transparent 1px),
		  linear-gradient(to bottom, rgba(30, 30, 30, .05) 1px, transparent 1px);
}

svg.railroad rect.railroad_canvas {
stroke-width: 0px;
fill: none;
}

svg.railroad path {
stroke-width: 3px;
stroke: black;
fill: none;
}

svg.railroad .debug {
stroke-width: 1px;
stroke: red;
}

svg.railroad text {
font: 14px monospace;
text-anchor: middle;
}

svg.railroad .nonterminal text {
font-weight: bold;
}

svg.railroad text.comment {
font: italic 12px monospace;
}

svg.railroad rect {
stroke-width: 3px;
stroke: black;
fill:hsl(-290, 70%, 90%);
}

svg.railroad g.labeledbox > rect {
stroke-width: 1px;
stroke: grey;
stroke-dasharray: 5px;
fill:rgba(90, 90, 150, .1);
}
</style>
<rect class="railroad_canvas" height="100%" width="100%"/>
<g class="verticalgrid">
<g class="sequence">
<path d=" M 10 34 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="54" y="39">
rule</text>
<g class="sequence">
<g class="optional">
<path d=" M 83 34 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 107 34 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="119" y="23"/>
<text x="141" y="39">
foo</text>
</g>
</g>
</g>
</g>
<path d=" M 73 34 h 10"/>
</g>
<path d=" M 209 34 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 34 h 10"/>
<path d=" M 199 34 h 10"/>
</g>
</g>
</svg>
